### DISCUSSION TITLE
How to manage the life cycle of installed Helm charts?

### DISCUSSION THREAD

**Discussion by otavio** [kube-hetzner/terraform-hcloud-kube-hetzner#316] [Q&A] [answered]
There are two prominent examples in the project which demand long-term maintenance such as:

1. Ingress controller
2. Longhorn operator

How would upgrades for those be applied in an existing cluster?

**Comment by mysticaltech** [comment, marked as answer]
@otavio Both have `HelmChart` and `HelmChartConfig` definitions, you can view them with `kubectl get helmchart(config) -A` (or just use `kubectl get crds` if I got the orthograph wrong).

You can upgrade them by applying a `HelmChartConfig`; see examples in the templates folder. And more on the matter here: https://rancher.com/docs/k3s/latest/en/helm/.

Also, it seems possible to directly upgrade with the `helm` cli, but GitOps principles would dictate the former method.

Lastly, we export a `kustomization_backup.yaml` and save it at the root of the folder; this will also include all the generated HelmCharts and HelmChartConfigs and can be used to upgrade too, but changing the chart versions. Please see the upgrade section in the Readme.

**Reply by otavio** [reply]
I am using https://fluxcd.io/, and we'd like to use it to provision all services into it. Seems like HelmChatConfig conflicts with this. Am I wrong?

**Reply by mysticaltech** [reply]
I am not that familiar with Flux @otavio; just know that it's close to ArgoCD,  but it could be the case, so you can disable the above add-ons to install them freshly with flux after the cluster is deployed! 

For ingresses:
```
enable_traefik = false
enable_nginx = false
```

For Longhorn:
```
enable_longhorn = false
extra_packages_to_install = ["open-iscsi", "nfs-client", "xfsprogs"]
```

The extra packages above are needed for Longhorn. I hope this helps! :)

**Reply by mysticaltech** [reply]
Our Longhorn config does additional steps like creating a `/var/longhorn` folder on the node or creating volumes and mounting those to `/var/longhorn` when the `longhorn_volume_size` attribute is present on the agent nodepool definition.

So if you want to take advantage of that, just keep `enable_longhorn = true`. The needed package will be installed automatically, and the proper node-level config will be done too. But on flux's side, you can start by deleting the HelmChart corresponding to Longhorn, leaving you with a fresh Longhorn-ready cluster!

Up to you, really, or you could have a separate TF module or Ansible script do the Longhorn prep work; that would work too.

**Reply by otavio** [reply]
We are preparing a testing deployment doing the installation using Helm and adding the extra packages as you suggested.

**Reply by mysticaltech** [reply]
Wonderful! If you manage to get this to work and are able to share, a PR adding to the examples folder, and a little note on the examples section in the readme, would be very much appreciated 🙏

**Reply by mrclrchtr** [reply]
An info for others, so they don't stumble over it like me: If you turn off longhorn the iscsid service is not activated anymore:
`inline = concat(var.enable_longhorn ? ["systemctl enable --now iscsid"] : [], [`

And
```
enable_traefik = false
enable_nginx = false
```
do not exist anymore.
Its now:
`ingress_controller = "none"`
